In [ ]:
#Import all of the necessary libraries needed
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [ ]:
#Read CSV file into a dataframe
df = pd.read_csv("/content/Lyftdataset.csv")

#Copy dataframe to a new variable to avoid any accidental changes in the original data
df_new = df.copy()

#Display the dataframe
display(df_new)

instant    dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0            1  01-01-11       1   0     1   0        0        6           0   
1            2  01-01-11       1   0     1   1        0        6           0   
2            3  01-01-11       1   0     1   2        0        6           0   
3            4  01-01-11       1   0     1   3        0        6           0   
4            5  01-01-11       1   0     1   4        0        6           0   
...        ...       ...     ...  ..   ...  ..      ...      ...         ...   
17374    17375  31-12-12       1   1    12  19        0        1           1   
17375    17376  31-12-12       1   1    12  20        0        1           1   
17376    17377  31-12-12       1   1    12  21        0        1           1   
17377    17378  31-12-12       1   1    12  22        0        1           1   
17378    17379  31-12-12       1   1    12  23        0        1           1   

       weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0               1  0.24  0.2879  0.81     0.0000       3          13   16  
1               1  0.22  0.2727  0.80     0.0000       8          32   40  
2               1  0.22  0.2727  0.80     0.0000       5          27   32  
3               1  0.24  0.2879  0.75     0.0000       3          10   13  
4               1  0.24  0.2879  0.75     0.0000       0           1    1  
...           ...   ...     ...   ...        ...     ...         ...  ...  
17374           2  0.26  0.2576  0.60     0.1642      11         108  119  
17375           2  0.26  0.2576  0.60     0.1642       8          81   89  
17376           1  0.26  0.2576  0.60     0.1642       7          83   90  
17377           1  0.26  0.2727  0.56     0.1343      13          48   61  
17378           1  0.26  0.2727  0.65     0.1343      12          37   49  

[17379 rows x 17 columns]

In [ ]:
#Print the type of the dataset
type(df_new)

pandas.core.frame.DataFrame

In [ ]:
#Display the number of rows and columns in the dataframe
df_new.shape

(17379, 17)

In [ ]:
#Display the columns in the dataset
df_new.keys()

Index(['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday',
       'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed',
       'casual', 'registered', 'cnt'],
      dtype='object')

In [ ]:
#Display information about the dataframe
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     17379 non-null  int64  
 1   dteday      17379 non-null  object 
 2   season      17379 non-null  int64  
 3   yr          17379 non-null  int64  
 4   mnth        17379 non-null  int64  
 5   hr          17379 non-null  int64  
 6   holiday     17379 non-null  int64  
 7   weekday     17379 non-null  int64  
 8   workingday  17379 non-null  int64  
 9   weathersit  17379 non-null  int64  
 10  temp        17379 non-null  float64
 11  atemp       17379 non-null  float64
 12  hum         17379 non-null  float64
 13  windspeed   17379 non-null  float64
 14  casual      17379 non-null  int64  
 15  registered  17379 non-null  int64  
 16  cnt         17379 non-null  int64  
dtypes: float64(4), int64(12), object(1)
memory usage: 2.3+ MB


In [ ]:
#Displays output of describe function and converting values that are displayed from scientific notation to normal numeric values
df_new.describe().apply(lambda s: s.apply('{0:.5f}'.format))

instant       season           yr         mnth           hr  \
count  17379.00000  17379.00000  17379.00000  17379.00000  17379.00000   
mean    8690.00000      2.50164      0.50256      6.53778     11.54675   
std     5017.02950      1.10692      0.50001      3.43878      6.91441   
min        1.00000      1.00000      0.00000      1.00000      0.00000   
25%     4345.50000      2.00000      0.00000      4.00000      6.00000   
50%     8690.00000      3.00000      1.00000      7.00000     12.00000   
75%    13034.50000      3.00000      1.00000     10.00000     18.00000   
max    17379.00000      4.00000      1.00000     12.00000     23.00000   

           holiday      weekday   workingday   weathersit         temp  \
count  17379.00000  17379.00000  17379.00000  17379.00000  17379.00000   
mean       0.02877      3.00368      0.68272      1.42528      0.49699   
std        0.16717      2.00577      0.46543      0.63936      0.19256   
min        0.00000      0.00000      0.00000      1.00000      0.02000   
25%        0.00000      1.00000      0.00000      1.00000      0.34000   
50%        0.00000      3.00000      1.00000      1.00000      0.50000   
75%        0.00000      5.00000      1.00000      2.00000      0.66000   
max        1.00000      6.00000      1.00000      4.00000      1.00000   

             atemp          hum    windspeed       casual   registered  \
count  17379.00000  17379.00000  17379.00000  17379.00000  17379.00000   
mean       0.47578      0.62723      0.19010     35.67622    153.78687   
std        0.17185      0.19293      0.12234     49.30503    151.35729   
min        0.00000      0.00000      0.00000      0.00000      0.00000   
25%        0.33330      0.48000      0.10450      4.00000     34.00000   
50%        0.48480      0.63000      0.19400     17.00000    115.00000   
75%        0.62120      0.78000      0.25370     48.00000    220.00000   
max        1.00000      1.00000      0.85070    367.00000    886.00000   

               cnt  
count  17379.00000  
mean     189.46309  
std      181.38760  
min        1.00000  
25%       40.00000  
50%      142.00000  
75%      281.00000  
max      977.00000

In [ ]:
#Check if any variable contains missing values
#It will return TRUE beside the column that contains missing values
df_new.isnull().any(axis=0)

instant       False
dteday        False
season        False
yr            False
mnth          False
hr            False
holiday       False
weekday       False
workingday    False
weathersit    False
temp          False
atemp         False
hum           False
windspeed     False
casual        False
registered    False
cnt           False
dtype: bool

In [ ]:
#Check if there are any duplicated rows
df_new.duplicated().sum()

0

In [ ]:
#Prepare new dataset by dropping "date and day" and "year" attributes.
##instant column is dropped as it is an row identifier.

df_new = df_new.drop(columns=['dteday','yr','instant'])
display(df_new)

season  mnth  hr  holiday  weekday  workingday  weathersit  temp  \
0           1     1   0        0        6           0           1  0.24   
1           1     1   1        0        6           0           1  0.22   
2           1     1   2        0        6           0           1  0.22   
3           1     1   3        0        6           0           1  0.24   
4           1     1   4        0        6           0           1  0.24   
...       ...   ...  ..      ...      ...         ...         ...   ...   
17374       1    12  19        0        1           1           2  0.26   
17375       1    12  20        0        1           1           2  0.26   
17376       1    12  21        0        1           1           1  0.26   
17377       1    12  22        0        1           1           1  0.26   
17378       1    12  23        0        1           1           1  0.26   

        atemp   hum  windspeed  casual  registered  cnt  
0      0.2879  0.81     0.0000       3          13   16  
1      0.2727  0.80     0.0000       8          32   40  
2      0.2727  0.80     0.0000       5          27   32  
3      0.2879  0.75     0.0000       3          10   13  
4      0.2879  0.75     0.0000       0           1    1  
...       ...   ...        ...     ...         ...  ...  
17374  0.2576  0.60     0.1642      11         108  119  
17375  0.2576  0.60     0.1642       8          81   89  
17376  0.2576  0.60     0.1642       7          83   90  
17377  0.2727  0.56     0.1343      13          48   61  
17378  0.2727  0.65     0.1343      12          37   49  

[17379 rows x 14 columns]

In [ ]:
#2 more attributes need to be dropped. Therefore a correlation matrix needs to be implemented to see the top 2 variables that are correlated to the target variable.
# this correlation matrix is being done before the train test split so that it would not affect the training and testing data.
#in this case, it can be seen that the top 2 variables that are closely correlated to the target variable are 'registered' and 'casual'. Therefore these variables need to be dropped.
df_new.corr()['cnt'].sort_values(ascending=False)

cnt           1.000000
registered    0.972151
casual        0.694564
temp          0.404772
atemp         0.400929
hr            0.394071
season        0.178056
mnth          0.120638
windspeed     0.093234
workingday    0.030284
weekday       0.026900
holiday      -0.030927
weathersit   -0.142426
hum          -0.322911
Name: cnt, dtype: float64

In [ ]:
#Drop "registered" and "casual" attributes

new_dataset = df_new.drop(columns=['registered','casual'])
display(new_dataset)

season  mnth  hr  holiday  weekday  workingday  weathersit  temp  \
0           1     1   0        0        6           0           1  0.24   
1           1     1   1        0        6           0           1  0.22   
2           1     1   2        0        6           0           1  0.22   
3           1     1   3        0        6           0           1  0.24   
4           1     1   4        0        6           0           1  0.24   
...       ...   ...  ..      ...      ...         ...         ...   ...   
17374       1    12  19        0        1           1           2  0.26   
17375       1    12  20        0        1           1           2  0.26   
17376       1    12  21        0        1           1           1  0.26   
17377       1    12  22        0        1           1           1  0.26   
17378       1    12  23        0        1           1           1  0.26   

        atemp   hum  windspeed  cnt  
0      0.2879  0.81     0.0000   16  
1      0.2727  0.80     0.0000   40  
2      0.2727  0.80     0.0000   32  
3      0.2879  0.75     0.0000   13  
4      0.2879  0.75     0.0000    1  
...       ...   ...        ...  ...  
17374  0.2576  0.60     0.1642  119  
17375  0.2576  0.60     0.1642   89  
17376  0.2576  0.60     0.1642   90  
17377  0.2727  0.56     0.1343   61  
17378  0.2727  0.65     0.1343   49  

[17379 rows x 12 columns]

In [ ]:
#Get independent variables for dataset X
x = new_dataset[['season','mnth','hr','holiday','weekday','workingday','weathersit','temp','atemp','hum','windspeed']]

#Get target variable for dataset Y
y = new_dataset[['cnt']]

display(x.head())
display(y.head())

season  mnth  hr  holiday  weekday  workingday  weathersit  temp   atemp  \
0       1     1   0        0        6           0           1  0.24  0.2879   
1       1     1   1        0        6           0           1  0.22  0.2727   
2       1     1   2        0        6           0           1  0.22  0.2727   
3       1     1   3        0        6           0           1  0.24  0.2879   
4       1     1   4        0        6           0           1  0.24  0.2879   

    hum  windspeed  
0  0.81        0.0  
1  0.80        0.0  
2  0.80        0.0  
3  0.75        0.0  
4  0.75        0.0

cnt
0   16
1   40
2   32
3   13
4    1

In [ ]:
#Split the data into 80:20 ratio.
X_train, X_test, Y_train, Y_test = train_test_split(x, y, test_size=0.2)

#doublecheck dimensions of training and testing data
print("X_train: ",X_train.shape)
print("X_test: ",X_test.shape)
print("Y_train: ",Y_train.shape)
print("Y_test: ",Y_test.shape)

X_train:  (13903, 11)
X_test:  (3476, 11)
Y_train:  (13903, 1)
Y_test:  (3476, 1)


In [ ]:
X_train.head()

season  mnth  hr  holiday  weekday  workingday  weathersit  temp  \
11679       2     5  20        0        0           0           1  0.60   
135         1     1  21        0        4           1           2  0.22   
14314       3     8  15        0        5           1           2  0.76   
3379        2     5  17        0        3           1           1  0.74   
11479       2     4  12        0        6           0           2  0.38   

        atemp   hum  windspeed  
11679  0.6061  0.64     0.1940  
135    0.2121  0.55     0.2239  
14314  0.6818  0.48     0.1343  
3379   0.6667  0.51     0.2239  
11479  0.3939  0.43     0.0000

In [ ]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13903 entries, 11679 to 8417
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   season      13903 non-null  int64  
 1   mnth        13903 non-null  int64  
 2   hr          13903 non-null  int64  
 3   holiday     13903 non-null  int64  
 4   weekday     13903 non-null  int64  
 5   workingday  13903 non-null  int64  
 6   weathersit  13903 non-null  int64  
 7   temp        13903 non-null  float64
 8   atemp       13903 non-null  float64
 9   hum         13903 non-null  float64
 10  windspeed   13903 non-null  float64
dtypes: float64(4), int64(7)
memory usage: 1.3 MB


In [ ]:
X_train.describe().apply(lambda s: s.apply('{0:.5f}'.format))

season         mnth           hr      holiday      weekday  \
count  13903.00000  13903.00000  13903.00000  13903.00000  13903.00000   
mean       2.49867      6.53003     11.52054      0.02906      2.99986   
std        1.10757      3.43379      6.92686      0.16798      2.00908   
min        1.00000      1.00000      0.00000      0.00000      0.00000   
25%        2.00000      4.00000      6.00000      0.00000      1.00000   
50%        3.00000      7.00000     12.00000      0.00000      3.00000   
75%        3.00000     10.00000     18.00000      0.00000      5.00000   
max        4.00000     12.00000     23.00000      1.00000      6.00000   

        workingday   weathersit         temp        atemp          hum  \
count  13903.00000  13903.00000  13903.00000  13903.00000  13903.00000   
mean       0.68043      1.42566      0.49722      0.47609      0.62710   
std        0.46633      0.64082      0.19240      0.17163      0.19333   
min        0.00000      1.00000      0.02000      0.00000      0.00000   
25%        0.00000      1.00000      0.34000      0.33330      0.48000   
50%        1.00000      1.00000      0.50000      0.48480      0.63000   
75%        1.00000      2.00000      0.66000      0.62120      0.78000   
max        1.00000      4.00000      1.00000      0.98480      1.00000   

         windspeed  
count  13903.00000  
mean       0.18963  
std        0.12197  
min        0.00000  
25%        0.10450  
50%        0.16420  
75%        0.25370  
max        0.85070

In [ ]:
X_test.head()

season  mnth  hr  holiday  weekday  workingday  weathersit  temp  \
6334        4     9  12        0        1           1           2  0.68   
8100        4    12   4        0        5           1           1  0.22   
3411        2     5   1        0        5           1           1  0.70   
8428        1    12  20        0        4           1           3  0.42   
15033       4     9  14        0        0           0           1  0.60   

        atemp   hum  windspeed  
6334   0.6364  0.79     0.1642  
8100   0.2727  0.80     0.0000  
3411   0.6515  0.58     0.2985  
8428   0.4242  0.71     0.0000  
15033  0.6212  0.38     0.1940

In [ ]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3476 entries, 6334 to 14400
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   season      3476 non-null   int64  
 1   mnth        3476 non-null   int64  
 2   hr          3476 non-null   int64  
 3   holiday     3476 non-null   int64  
 4   weekday     3476 non-null   int64  
 5   workingday  3476 non-null   int64  
 6   weathersit  3476 non-null   int64  
 7   temp        3476 non-null   float64
 8   atemp       3476 non-null   float64
 9   hum         3476 non-null   float64
 10  windspeed   3476 non-null   float64
dtypes: float64(4), int64(7)
memory usage: 325.9 KB


In [ ]:
X_test.describe().apply(lambda s: s.apply('{0:.5f}'.format))

season        mnth          hr     holiday     weekday  workingday  \
count  3476.00000  3476.00000  3476.00000  3476.00000  3476.00000  3476.00000   
mean      2.51352     6.56876    11.65161     0.02762     3.01899     0.69189   
std       1.10439     3.45897     6.86434     0.16390     1.99270     0.46178   
min       1.00000     1.00000     0.00000     0.00000     0.00000     0.00000   
25%       2.00000     4.00000     6.00000     0.00000     1.00000     0.00000   
50%       3.00000     7.00000    12.00000     0.00000     3.00000     1.00000   
75%       3.00000    10.00000    17.00000     0.00000     5.00000     1.00000   
max       4.00000    12.00000    23.00000     1.00000     6.00000     1.00000   

       weathersit        temp       atemp         hum   windspeed  
count  3476.00000  3476.00000  3476.00000  3476.00000  3476.00000  
mean      1.42376     0.49607     0.47451     0.62772     0.19198  
std       0.63355     0.19318     0.17276     0.19136     0.12382  
min       1.00000     0.02000     0.00000     0.00000     0.00000  
25%       1.00000     0.34000     0.33330     0.47000     0.10450  
50%       1.00000     0.50000     0.48480     0.63000     0.19400  
75%       2.00000     0.66000     0.62120     0.78000     0.25370  
max       3.00000     0.96000     1.00000     1.00000     0.68660

In [ ]:
Y_train.head()

cnt
11679  276
135     48
14314  350
3379   601
11479  381

In [ ]:
Y_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13903 entries, 11679 to 8417
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   cnt     13903 non-null  int64
dtypes: int64(1)
memory usage: 217.2 KB


In [ ]:
Y_train.describe().apply(lambda s: s.apply('{0:.5f}'.format))

cnt
count  13903.00000
mean     189.84421
std      181.26485
min        1.00000
25%       40.00000
50%      143.00000
75%      282.50000
max      977.00000

In [ ]:
Y_test.head()

cnt
6334   209
8100     2
3411    50
8428    33
15033  666

In [ ]:
Y_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3476 entries, 6334 to 14400
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   cnt     3476 non-null   int64
dtypes: int64(1)
memory usage: 54.3 KB


In [ ]:
Y_test.describe().apply(lambda s: s.apply('{0:.5f}'.format))

cnt
count  3476.00000
mean    187.93872
std     181.89589
min       1.00000
25%      40.00000
50%     140.00000
75%     274.00000
max     948.00000

In [ ]:
xa = X_train
xa['cnt'] = Y_train
xa.head()

season  mnth  hr  holiday  weekday  workingday  weathersit  temp  \
11679       2     5  20        0        0           0           1  0.60   
135         1     1  21        0        4           1           2  0.22   
14314       3     8  15        0        5           1           2  0.76   
3379        2     5  17        0        3           1           1  0.74   
11479       2     4  12        0        6           0           2  0.38   

        atemp   hum  windspeed  cnt  
11679  0.6061  0.64     0.1940  276  
135    0.2121  0.55     0.2239   48  
14314  0.6818  0.48     0.1343  350  
3379   0.6667  0.51     0.2239  601  
11479  0.3939  0.43     0.0000  381

In [ ]:
corr = xa.corr()
corr.style.background_gradient(cmap='coolwarm')

In [ ]:
xa.corr()['cnt'].sort_values(ascending=False)

cnt           1.000000
temp          0.406322
atemp         0.402281
hr            0.397351
season        0.178728
mnth          0.119997
windspeed     0.092551
workingday    0.029524
weekday       0.027602
holiday      -0.033138
weathersit   -0.146176
hum          -0.324183
Name: cnt, dtype: float64

In [ ]:
#Select the 4 parameters that are the closest to 1
sel = ['temp','atemp','hr','season']
X_new = X_train[sel]
X_new_test = X_test[sel]
X_new.head()

temp   atemp  hr  season
11679  0.60  0.6061  20       2
135    0.22  0.2121  21       1
14314  0.76  0.6818  15       3
3379   0.74  0.6667  17       2
11479  0.38  0.3939  12       2

In [ ]:
Y_train.head()

cnt
11679  276
135     48
14314  350
3379   601
11479  381

In [ ]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler
scaler = StandardScaler() # scaler = MinMaxScaler()
X_new1 = scaler.fit_transform(X_new)
X_new_test1 = scaler.transform(X_new_test)

scaler_y = StandardScaler()
Y_train1 = scaler_y.fit_transform(Y_train)
Y_test1 = scaler_y.transform(Y_test)

In [ ]:
X_new1 = pd.DataFrame(X_new1, columns = sel)
X_new_test1 = pd.DataFrame(X_new_test1, columns = sel)

Y_train1 = pd.DataFrame(Y_train1, columns=['Y'])
Y_test1 = pd.DataFrame(Y_test1, columns=['Y'])
X_new1.head()

temp     atemp        hr    season
0  0.534224  0.757533  1.224186 -0.450254
1 -1.440849 -1.538244  1.368557 -1.353164
2  1.365833  1.198626  0.502333  0.452657
3  1.261882  1.110640  0.791074 -0.450254
4 -0.609239 -0.478924  0.069221 -0.450254

In [ ]:
X_new1.describe()

temp         atemp            hr        season
count  1.390300e+04  1.390300e+04  1.390300e+04  1.390300e+04
mean  -5.395638e-16 -1.075806e-16 -1.052807e-16  1.650761e-16
std    1.000036e+00  1.000036e+00  1.000036e+00  1.000036e+00
min   -2.480361e+00 -2.774118e+00 -1.663228e+00 -1.353164e+00
25%   -8.171418e-01 -8.320304e-01 -7.970034e-01 -4.502537e-01
50%    1.446778e-02  5.073682e-02  6.922067e-02  4.526566e-01
75%    8.460774e-01  8.455187e-01  9.354448e-01  4.526566e-01
max    2.613248e+00  2.964160e+00  1.657298e+00  1.355567e+00

In [ ]:
X_new.head()

temp   atemp  hr  season
11679  0.60  0.6061  20       2
135    0.22  0.2121  21       1
14314  0.76  0.6818  15       3
3379   0.74  0.6667  17       2
11479  0.38  0.3939  12       2

In [ ]:
Y_train.head()

cnt
11679  276
135     48
14314  350
3379   601
11479  381

In [ ]:
Y_train1.head()

Y
0  0.475320
1 -0.782553
2  0.883578
3  2.268342
4  1.054604

In [ ]:
from sklearn.linear_model import SGDRegressor

regr = SGDRegressor(max_iter = 10000)
regr1 = SGDRegressor()


In [ ]:
#Create linear regression object without scaling
regr = LinearRegression() #making an instance

#Train the model using the training sets
regr.fit(X_new, Y_train)

#Make predictions using the testing set
Y_pred = regr.predict(X_new_test)

#The coefficients
print('Coefficients: \n', regr.coef_)
print('Intercept or Bias: \n', regr.intercept_)

#The mean squared error
print("Mean squared error: %.2f" % mean_squared_error(Y_test, Y_pred))

#The Mean absolute error
print('Mean absolute error: %.2f' % mean_absolute_error(Y_test, Y_pred))

#Explained variance score (R2 Score): 1 is perfect prediction
print('R2 score: %.2f' % r2_score(Y_test, Y_pred))

Coefficients: 
 [[279.6638973   38.119747     9.16730189  12.43054946]]
Intercept or Bias: 
 [-104.02986289]
Mean squared error: 23899.74
Mean absolute error: 113.14
R2 score: 0.28


In [ ]:
# Create linear regression object with scaling

# Train the model using the training sets
regr1.fit(X_new1, Y_train1)

# Make predictions using the testing set
Y_pred1 = regr1.predict(X_new_test1)

# The coefficients
print('Coefficients: \n', regr1.coef_)
print ('Intercept: ', regr1.intercept_)
# The mean squared error
print("Mean squared error: %.2f"
      % mean_squared_error(Y_test1, Y_pred1))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % r2_score(Y_test1, Y_pred1))
#Explained variance score (R2 Score): 1 is perfect prediction
print('R2 score: %.2f' % r2_score(Y_test1, Y_pred1))

Coefficients: 
 [ 0.12177035  0.1816062   0.26987815  0.11358493 -0.0076728   0.02105144
  0.02611341  0.03658595 -0.01283877 -0.01122963 -0.2488326 ]
Intercept:  [0.00959344]
Mean squared error: 0.67
Variance score: 0.33
R2 score: 0.33


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
# Create linear regression object with unscaled dataset
regr2 = LinearRegression()

# Train the model using the training sets
regr2.fit(X_new, Y_train)

# Make predictions using the testing set
Y_pred2 = regr2.predict(X_new_test)

# The coefficients
print('Coefficients: \n', regr2.coef_)
# The mean squared error
print("Mean squared error: %.2f"
      % mean_squared_error(Y_test, Y_pred2))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % r2_score(Y_test, Y_pred2))
#Explained variance score (R2 Score): 1 is perfect prediction
print('R2 score: %.2f' % r2_score(Y_test, Y_pred2))

Coefficients: 
 [[ 9.91414610e+01  2.18594556e+02  7.45397128e+00  1.93382428e+01
   9.52287136e-02  2.61066298e+01  3.13465725e+00  1.72042968e+00
  -2.24115109e+01 -9.85598933e-01 -2.20795496e+02]]
Mean squared error: 21985.65
Variance score: 0.34
R2 score: 0.34


In [ ]:
# Create linear regression object with scaled data
regr3 = LinearRegression()

# Train the model using the training sets
regr3.fit(X_new1, Y_train1)

# Make predictions using the testing set
Y_pred3 = regr3.predict(X_new_test1)

# The coefficients
print('Coefficients: \n', regr3.coef_)
# The mean squared error
print("Mean squared error: %.2f"
      % mean_squared_error(Y_test1, Y_pred3))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % r2_score(Y_test1, Y_pred3))
#Explained variance score (R2 Score): 1 is perfect prediction
print('R2 score: %.2f' % r2_score(Y_test1, Y_pred3))

Coefficients: 
 [[ 0.10523443  0.20697019  0.2848464   0.11816107  0.00180396  0.01756634
   0.00806431  0.01906867 -0.02076855 -0.00348438 -0.23548818]]
Mean squared error: 0.67
Variance score: 0.34
R2 score: 0.34


#Include all features of new dataset and perform linear regression

In [ ]:
#Select all parameters
sel = ['temp','atemp','hr','season','mnth','windspeed','workingday','weekday','holiday','weathersit','hum']
X_new = X_train[sel]
X_new_test = X_test[sel]
X_new.head()

temp   atemp  hr  season  mnth  windspeed  workingday  weekday  \
11679  0.60  0.6061  20       2     5     0.1940           0        0   
135    0.22  0.2121  21       1     1     0.2239           1        4   
14314  0.76  0.6818  15       3     8     0.1343           1        5   
3379   0.74  0.6667  17       2     5     0.2239           1        3   
11479  0.38  0.3939  12       2     4     0.0000           0        6   

       holiday  weathersit   hum  
11679        0           1  0.64  
135          0           2  0.55  
14314        0           2  0.48  
3379         0           1  0.51  
11479        0           2  0.43

In [ ]:
Y_train.head()

cnt
11679  276
135     48
14314  350
3379   601
11479  381

In [ ]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler
scaler = StandardScaler() # scaler = MinMaxScaler()
X_new1 = scaler.fit_transform(X_new)
X_new_test1 = scaler.transform(X_new_test)

scaler_y = StandardScaler()
Y_train1 = scaler_y.fit_transform(Y_train)
Y_test1 = scaler_y.transform(Y_test)

In [ ]:
X_new1 = pd.DataFrame(X_new1, columns = sel)
X_new_test1 = pd.DataFrame(X_new_test1, columns = sel)

Y_train1 = pd.DataFrame(Y_train1, columns=['Y'])
Y_test1 = pd.DataFrame(Y_test1, columns=['Y'])
X_new1.head()

temp     atemp        hr    season      mnth  windspeed  workingday  \
0  0.534224  0.757533  1.224186 -0.450254 -0.445597   0.035851   -1.459175   
1 -1.440849 -1.538244  1.368557 -1.353164 -1.610532   0.281007    0.685319   
2  1.365833  1.198626  0.502333  0.452657  0.428105  -0.453642    0.685319   
3  1.261882  1.110640  0.791074 -0.450254 -0.445597   0.281007    0.685319   
4 -0.609239 -0.478924  0.069221 -0.450254 -0.736830  -1.554795   -1.459175   

    weekday   holiday  weathersit       hum  
0 -1.493204 -0.172998   -0.664268  0.066703  
1  0.497830 -0.172998    0.896280 -0.398846  
2  0.995589 -0.172998    0.896280 -0.760940  
3  0.000072 -0.172998   -0.664268 -0.605757  
4  1.493347 -0.172998    0.896280 -1.019578

In [ ]:
X_new1.describe()

temp         atemp            hr        season          mnth  \
count  1.390300e+04  1.390300e+04  1.390300e+04  1.390300e+04  1.390300e+04   
mean  -5.395638e-16 -1.075806e-16 -1.052807e-16  1.650761e-16  6.848358e-17   
std    1.000036e+00  1.000036e+00  1.000036e+00  1.000036e+00  1.000036e+00   
min   -2.480361e+00 -2.774118e+00 -1.663228e+00 -1.353164e+00 -1.610532e+00   
25%   -8.171418e-01 -8.320304e-01 -7.970034e-01 -4.502537e-01 -7.368304e-01   
50%    1.446778e-02  5.073682e-02  6.922067e-02  4.526566e-01  1.368714e-01   
75%    8.460774e-01  8.455187e-01  9.354448e-01  4.526566e-01  1.010573e+00   
max    2.613248e+00  2.964160e+00  1.657298e+00  1.355567e+00  1.593041e+00   

          windspeed    workingday       weekday       holiday    weathersit  \
count  1.390300e+04  1.390300e+04  1.390300e+04  1.390300e+04  1.390300e+04   
mean   1.155022e-16  1.073250e-17 -1.052807e-16  5.570680e-17 -3.884144e-17   
std    1.000036e+00  1.000036e+00  1.000036e+00  1.000036e+00  1.000036e+00   
min   -1.554795e+00 -1.459175e+00 -1.493204e+00 -1.729975e-01 -6.642684e-01   
25%   -6.979780e-01 -1.459175e+00 -9.954453e-01 -1.729975e-01 -6.642684e-01   
50%   -2.084856e-01  6.853187e-01  7.160447e-05 -1.729975e-01 -6.642684e-01   
75%    5.253430e-01  6.853187e-01  9.955885e-01 -1.729975e-01  8.962797e-01   
max    5.420267e+00  6.853187e-01  1.493347e+00  5.780430e+00  4.017376e+00   

                hum  
count  1.390300e+04  
mean   3.430568e-16  
std    1.000036e+00  
min   -3.243869e+00  
25%   -7.609400e-01  
50%    1.497547e-02  
75%    7.908909e-01  
max    1.928900e+00

In [ ]:
X_new.head()

temp   atemp  hr  season  mnth  windspeed  workingday  weekday  \
11679  0.60  0.6061  20       2     5     0.1940           0        0   
135    0.22  0.2121  21       1     1     0.2239           1        4   
14314  0.76  0.6818  15       3     8     0.1343           1        5   
3379   0.74  0.6667  17       2     5     0.2239           1        3   
11479  0.38  0.3939  12       2     4     0.0000           0        6   

       holiday  weathersit   hum  
11679        0           1  0.64  
135          0           2  0.55  
14314        0           2  0.48  
3379         0           1  0.51  
11479        0           2  0.43

In [ ]:
Y_train.head()

cnt
11679  276
135     48
14314  350
3379   601
11479  381

In [ ]:
Y_train1.head()

Y
0  0.475320
1 -0.782553
2  0.883578
3  2.268342
4  1.054604

In [ ]:
from sklearn.linear_model import SGDRegressor

regr = SGDRegressor(max_iter = 10000)
regr1 = SGDRegressor()


In [ ]:
#Create linear regression object without scaling
regr = LinearRegression() #making an instance

#Train the model using the training sets
regr.fit(X_new, Y_train)

#Make predictions using the testing set
Y_pred = regr.predict(X_new_test)

#The coefficients
print('Coefficients: \n', regr.coef_)
print('Intercept or Bias: \n', regr.intercept_)

#The mean squared error
print("Mean squared error: %.2f" % mean_squared_error(Y_test, Y_pred))

#The Mean absolute error
print('Mean absolute error: %.2f' % mean_absolute_error(Y_test, Y_pred))

#Explained variance score (R2 Score): 1 is perfect prediction
print('R2 score: %.2f' % r2_score(Y_test, Y_pred))

Coefficients: 
 [[ 9.91414610e+01  2.18594556e+02  7.45397128e+00  1.93382428e+01
   9.52287136e-02  2.61066298e+01  3.13465725e+00  1.72042968e+00
  -2.24115109e+01 -9.85598933e-01 -2.20795496e+02]]
Intercept or Bias: 
 [29.93657382]
Mean squared error: 21985.65
Mean absolute error: 108.06
R2 score: 0.34


In [ ]:
# Create linear regression object with scaling

# Train the model using the training sets
regr1.fit(X_new1, Y_train1)

# Make predictions using the testing set
Y_pred1 = regr1.predict(X_new_test1)

# The coefficients
print('Coefficients: \n', regr1.coef_)
print ('Intercept: ', regr1.intercept_)
# The mean squared error
print("Mean squared error: %.2f"
      % mean_squared_error(Y_test1, Y_pred1))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % r2_score(Y_test1, Y_pred1))
#Explained variance score (R2 Score): 1 is perfect prediction
print('R2 score: %.2f' % r2_score(Y_test1, Y_pred1))

Coefficients: 
 [ 0.12198391  0.17842099  0.27307745  0.11220938  0.00550871  0.02317305
  0.002561    0.02417575 -0.00884366 -0.00732719 -0.21961826]
Intercept:  [-0.02185632]
Mean squared error: 0.67
Variance score: 0.33
R2 score: 0.33


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
# Create linear regression object with unscaled dataset
regr2 = LinearRegression()

# Train the model using the training sets
regr2.fit(X_new, Y_train)

# Make predictions using the testing set
Y_pred2 = regr2.predict(X_new_test)

# The coefficients
print('Coefficients: \n', regr2.coef_)
# The mean squared error
print("Mean squared error: %.2f"
      % mean_squared_error(Y_test, Y_pred2))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % r2_score(Y_test, Y_pred2))
#Explained variance score (R2 Score): 1 is perfect prediction
print('R2 score: %.2f' % r2_score(Y_test, Y_pred2))

Coefficients: 
 [[ 9.91414610e+01  2.18594556e+02  7.45397128e+00  1.93382428e+01
   9.52287136e-02  2.61066298e+01  3.13465725e+00  1.72042968e+00
  -2.24115109e+01 -9.85598933e-01 -2.20795496e+02]]
Mean squared error: 21985.65
Variance score: 0.34
R2 score: 0.34


In [ ]:
# Create linear regression object with scaled data
regr3 = LinearRegression()

# Train the model using the training sets
regr3.fit(X_new1, Y_train1)

# Make predictions using the testing set
Y_pred3 = regr3.predict(X_new_test1)

# The coefficients
print('Coefficients: \n', regr3.coef_)
# The mean squared error
print("Mean squared error: %.2f"
      % mean_squared_error(Y_test1, Y_pred3))
# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % r2_score(Y_test1, Y_pred3))
#Explained variance score (R2 Score): 1 is perfect prediction
print('R2 score: %.2f' % r2_score(Y_test1, Y_pred3))

Coefficients: 
 [[ 0.10523443  0.20697019  0.2848464   0.11816107  0.00180396  0.01756634
   0.00806431  0.01906867 -0.02076855 -0.00348438 -0.23548818]]
Mean squared error: 0.67
Variance score: 0.34
R2 score: 0.34


Linear Regression Model based on 4 parameters.

Without Scaling:
Mean squared error: 23899.74
Mean absolute error: 113.14
R2 score: 0.28

With unscaled dataset:
Mean squared error: 21985.65
Variance score: 0.34
R2 score: 0.34

With Scaling:
Mean squared error: 0.67
Variance score: 0.33
R2 score: 0.33

With scaled dataset:
Mean squared error: 0.67
Variance score: 0.34
R2 score: 0.34


Linear Regression Model with all parameters.

Without Scaling:
Mean squared error: 21985.65
Mean absolute error: 108.06
R2 score: 0.34


With unscaled dataset:
Mean squared error: 21985.65
Variance score: 0.34
R2 score: 0.34


With Scaling:
Mean squared error: 0.67
Variance score: 0.33
R2 score: 0.33


With scaled dataset:
Mean squared error: 0.67
Variance score: 0.34
R2 score: 0.34



Based on the following results, The model that has 4 parameters is slightly better than the model with all parameters. Therefore, I would choose the model with 4 parameters. Even if the results are slightly better, I would still prefer that model. In both models, scaled data do not make much of an impact.